Lab3

Andrew174194 opened the pull request:
Pull request for lab 3
by Andrey Khoroshavin (kill team, CS-01 academic group)

Cre-eD committed:
Init

Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

Cre-eD committed:
Init

Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

Cre-eD committed:
Update

Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

Andrew174194 committed:
Transfer from Lab1

Copy useful files from Lab1

Andrew174194 committed:
Lab 2

- Created Dockerfile and DOCKER.md
- Edited README with new Docker info
- Changed main.py for correct work from Docker
- Deleted prepare.sh due to it's inoperability

Andrew174194 committed:
Format fixes

Fixes for MarkDown syntax

Andrew174194 committed:
Squashed commit of the following:

commit dc3c1e4df1e7b43810c03877706787e3dfad038c
Author: Dmitriy Creed <creed@soramitsu.co.jp>
Date:   Mon Sep 12 22:50:45 2022 +0300

    lab3

    Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

commit 2980cf03b6e0c8588cc1bc76ee733f54ec14ffa3
Author: Dmitriy Creed <creed@soramitsu.co.jp>
Date:   Mon Sep 12 21:59:19 2022 +0300

    lab3

    Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

commit 79f2d05c12c88f47c38bba1202905c6e24b9c6ba
Author: Dmitriy Creed <creed@soramitsu.co.jp>
Date:   Tue Aug 30 01:14:51 2022 +0300

    Update

    Signed-off-by: Dmitriy Creed <creed@soramitsu.co.jp>

Andrew174194 committed:
Lab 3

Added Unit tests, CI pipeline & Documentation

Andrew174194 committed:
Fix CI pipeline

DOCtorActoAntohich commented:
since you're trying to merge into an an empty branch where these files don't exist, it counts them all as "new" ![](https://steamcommunity-a.akamaihd.net/economy/emoticon/:ohh_yeah:)

DOCtorActoAntohich commented on app_python/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM python:3.9-alpine
+EXPOSE 5000
+COPY app_python/main.py app_python/requirements.txt app_python/test.py ./app/
+RUN cd ./app/ && pip install -r requirements.txt && adduser -D app_user && chown -R root ./ && chmod -R 755 ./
```
check out `WORKDIR` ![](https://steamcommunity-a.akamaihd.net/economy/emoticon/:ohh_yeah:)
no problem when there's only one `cd` but when there's more, it may get confusing :trollface:

DOCtorActoAntohich commented on app_python/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+COPY app_python/main.py app_python/requirements.txt app_python/test.py ./app/
+RUN cd ./app/ && pip install -r requirements.txt && adduser -D app_user && chown -R root ./ && chmod -R 755 ./
+USER app_user
+CMD python ./app/main.py && python ./app/test.py
```
do you really need to run tests after you close main app? :trollface: 

not sure though, but i thought tests should be run separately before the main app, and shouldn't be touched in production ![](https://steamcommunity-a.akamaihd.net/economy/emoticon/:ohh_yeah:)

smthngslv replied:
app is not working? TEST IT!

ntdesmond replied:
Two teas to the gentleman above.

It seems that testing the code would just be a waste of time, if done during the Docker build process.
E.g. if you do it in CI, you can test the code before reaching docker-related steps.

DOCtorActoAntohich reviewed (approved):
looks okay... ish :trollface: 

mayhaps i'd reorganize stuff ![](https://steamcommunity-a.akamaihd.net/economy/emoticon/:ohh_yeah:) because `app_python` is getting kinda cluttered.

sadly, i have no idea how `unittest` works, but i hope it should allow you to separate tests and the main app. also you can move docker-related files out of this folder too.

but like they say :trollface: "if it works - don't touch it" ![](https://steamcommunity-a.akamaihd.net/economy/emoticon/:ohh_yeah:)

ntdesmond commented on .github/workflows/main.yaml:
```diff
@@ -0,0 +1,57 @@
+          file: ./app_python/Dockerfile
+          builder: ${{ steps.buildx.outputs.name }}
+          push: true
+          tags: ${{ secrets.DOCKER_HUB_USERNAME }}/msc_time_py:lab
```
better than `latest`, but still a possible security issue — anyone might open a PR and get their code pushed to docker hub

ntdesmond commented on .github/workflows/main.yaml:
```diff
@@ -0,0 +1,57 @@
+          push: true
+          tags: ${{ secrets.DOCKER_HUB_USERNAME }}/msc_time_py:lab
+          cache-from: type=registry,ref=${{ secrets.DOCKER_HUB_USERNAME }}/msc_time_py:buildcache
+          cache-to: type=registry,ref=${{ secrets.DOCKER_HUB_USERNAME }}/msc_time_py:buildcache,mode=max
```
Overall good CI but consider splitting Docker steps from the others. Check out how to [share the built code between the jobs](https://levelup.gitconnected.com/github-actions-how-to-share-data-between-jobs-fc1547defc3e).

ntdesmond commented on app_python/CI.md:
```diff
@@ -0,0 +1,5 @@
+# Best practicies for Docker Hub in CI workflow
+1. Do unit testing inside Socker container to avoid errors in containers
```
SOCKER

It's probably better to do testing before building a Docker container, but it's up to you.

smthngslv commented on .github/workflows/main.yaml:
```diff
@@ -0,0 +1,57 @@
+      - name: Lint with flake8
+        run: |
+          # stop the build if there are Python syntax errors or undefined names
+          flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
```
move config to file

smthngslv commented:
I like class based test case, GJ

ntdesmond commented on app_python/PYTHON.md:
```diff
@@ -0,0 +1,62 @@
+# Best applied python practices (and how I used them)
+
+1) **Use virtual environment to run product**
```
Thanks for eliminating `<br>`s but WHY `)`

molberte replied:
@ntdesmond dushnila

ntdesmond commented on app_python/test.py:
```diff
@@ -0,0 +1,26 @@
+        try:
+            self.webpage = requests.get("http://localhost:5000")    
+        except requests.ConnectionError:
+            raise unittest.SkipTest("Service is offline")
```
Skip if not running the server? Don't really get it lol

ntdesmond reviewed (approved):
# e

molberte commented:
wow you literally named you commits by fixes you applied. i'm impressed!

Andrew174194 committed:
Fixes for Lab3

- Rework GitHub Actions: tag testing containers with branch name & add workflow for RP to master
- Move info .md's to DOCS folder to keep app_python/ clean
- Fix unit-testing

Andrew174194 committed:
Merge branch 'innopolis-devops:lab3' into lab3

Andrew174194 committed:
Little fixes for CI
